### ISSUE TITLE
libpng12 not available in Debian 9 

### ISSUE THREAD

**Issue by alugarius** [pioneerspacesim/pioneer#3959]
I wanted to play Pioneer and i saw that it depends on an outdated libpng, i have libpng16 installed.
please update to a newer version, installing libpng12-0 is not supported, the sourcecode for libpng12 is hard to find. 
Users who are new to Linux, and me- have problems using your game and rate it as broken.

i am a regular Linux user, i want to help but i am not a programmer, so i promise donating €50 or more if that problem gets solved.
Also doing Promotion

**Comment by alugarius** [comment]
i have a depressive phase, sorry

**Comment by fluffyfreak** [comment]
I think we usually link against libpng15 or 16 already. However I'm not the Linux maintainer!
@impaktor  / @laarmen do you guys have any idea if we ask for a specific libpng version? I can't see that we do in the `configure.ac` or makefiles.

**Comment by laarmen** [comment]
We use the system one. The 16 is an ABI version, not an API, and it looks
like the good folks at libpng don't have any remorse in breaking ABI. This
is why Linux is a PITA for proprietary development...

Le lun. 3 avr. 2017 à 16:09, Andrew Copland <notifications@github.com> a
écrit :

**Comment by alugarius** [comment]
@laarmen 
sorry, i have no idea what a ABI is

**Comment by laarmen** [comment]
No problem. ABI is Application Binary Interface, whereas API stands for
Application Programming Interface. We as developers program against a
specific API, but the compiled programs are in effect built against a
specific ABI. There isn't necessarily a link between API compatibility and
ABI compat.

In the linux world, since we have access to most of the sources, and can
thus recompile our software, developers tended (or whatever the pas tense
of tend is) not to care much about ABI stability : you can always
recompile, right? This mindset is slowly evolving, though, which is good
news.

**Comment by pgimeno** [comment]
I think the OP means that the binary download for Linux dynamically links against libpng12. I was confused too when he brought it up on IRC.

```
$ ldd pioneer | grep libpng
	libpng12.so.0 => /lib/i386-linux-gnu/libpng12.so.0 (0xb74e5000)
```

**Comment by alugarius** [comment]
@pgimeno
./pioneer: /usr/lib/libpng12.so.0: no version information available (required by ./pioneer)
        libpng12.so.0 => /usr/lib/libpng12.so.0 (0x00007f7eb2294000)
        libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f7eb0b7e000)

**Comment by alugarius** [comment]
Someone thought about Appimage ? 
http://appimage.org/

**Comment by mike-f1** [comment]
@alugarius, 
1) could you check if you are on a 64 or a 32 bit environment? (...I suggest to use uname -a)
2) could you check if you have libpng12 installed? ...just because I don't understand from where 

> libpng12.so.0 => /usr/lib/libpng12.so.0 (0x00007f7eb2294000)

this line come... are this an output from a command?

Here a link to a page where I think (...I think!) you can download libpng12
for jessie (Ubuntu 8, right?). Ubuntu 9 is, AFAIK, under "test" phase...

https://packages.debian.org/jessie/libpng12-0

Could you paste the link from where you download this game?

...Lot of questions, but I hope could help dev team :)

Bye

**Comment by fluffyfreak** [comment]
@robn looks like we link against libpng-12 for x86 and libpng-16 for x86_64 on the build machine?

**Comment by alugarius** [comment]
I use 64 bit (Debian 9) and i copy pasted my libpng12 (from debian 8)

ubuntu 9 is from 2009 so it might not be in a "test" phase anymore, 

 Debian 9 rejects (automatically removes libpng12 if installed via apt)

and where else should i download your game as from your homepage?

@fluffyfreak  and no, the offical 64 bit is linked against libpng16 and libpng12 (witch is kind of hell for a Linux user)

**Comment by alugarius** [comment]
i didnt wanted to look like this... i am a mardown noob

**Comment by LuccoJ** [comment]
Actually I can confirm this. I am running 64-bit Intel Debian Stretch (testing), and had to install libpng12 from the Jessie (stable) repositories before I could run the latest (March) binary build offered to me on the website. I think that was the case with past builds, too, but I had forgotten about the requirement.

**Comment by alugarius** [comment]
Any progress here?

**Comment by laarmen** [comment]
Pinging @robn ?

**Comment by alugarius** [comment]
I am asking because Debian is now stable.... And Debian 8 will go away (which is the platform that pioneer is working on)

**Comment by alugarius** [comment]
*cough *....

**Comment by Talkless** [comment]
Discovered this on Ubuntu 17.10 daily and on Debian Testing.

Could it be linked statically to live through this transition?

**Comment by laarmen** [comment]
The thing is, our build box is owned and maintained by @robn, who doesn't
seem to have much time available for Pioneer except for the odd "run the
build-and-upload" script every few months. Which is fine by me, as it
allows for us to have a Windows build easily, and I thank him for that.
The issue here is that this is not going to be solved with a mere code
patch, you want a fix, you're going to need to provide infrastructure, as I
don't think it is fair of us to bully @robn into upgrading it (admittedly
old-old-stable) Debian box just for the sake of Pioneer. And that's
assuming that's even an option on the technical side.

To sum up : we're happy to upload whatever fixed binaries you provide us.
We can't do much more though.

**Comment by Talkless** [comment]
Is this `build-and-upload` script available? I could test on Wheezy virtual machine to make it link libpng (and libz. I believe libpng needs) statically.

**Comment by impaktor** [comment]
the buildscripts that are available are in scripts/ folder of our source tree.

**Comment by monolifed** [comment]
It has been nearly one year and still no solution? You can probably disable linux builds or place a warning on the download page notifying that it won't run on up-to-date linux distros.

**Comment by nozmajner** [comment]
It runs on current Mint without hassle.

**Comment by impaktor** [comment]
@monolifed Pretty much the entire dev team is on GNU/Linux, and in addition a lot of people use the binary build. It works. However, making a pre-built binary work on  _every_ linux distribution is another thing, but please don't assume it doesn't work on GNU/Linux _at all_ just because it doesn't happen to work on your particular version of your particular distribution.

**Comment by monolifed** [comment]
If you call Debian and Ubuntu, my particular version of my particular distribution, you can close the issue as well. I can always build it from the source, so this is not an issue for me either.
But I still think there needs to be a warning or a work around for those people using, their particular space sim on their particular version of their particular distributions

**Comment by nozmajner** [comment]
Mint is practically Ubuntu, and Pioneer runs here without a hitch, I didn't even had to install anything extra, just extracted and worked right away.

**Comment by impaktor** [comment]
Right, and Pioneer is built on a Debian machine, so will work by default on that version of Debian.

**Comment by monolifed** [comment]
Which Debian version?
https://packages.debian.org/search?keywords=libpng12-0
wheezy(oldoldstable), jessie(oldstable) and sid(unstable) has it. However stretch(current) and buster(development) do not.
For Ubuntu,
https://packages.ubuntu.com/search?keywords=libpng12-0
old and current LTS have it. However current version does not.

Thus, as far as I can tell, current versions of ubuntu and debian do not have it. But I have to admit using libpng16 will break it for some other folk:
https://packages.ubuntu.com/search?keywords=libpng16-16
https://packages.debian.org/search?keywords=libpng16-16

**Comment by Hoimar** [comment]
I wanted to add a solution for users that don't have the `libpng12` package in their default repo. So any *buntu after 16.04 like the upcoming LTS 18.04 which I'm using :smiley:. But this should work for any Debian-like Linux distro.
Anyway, after having downloaded the default Linux64 build from pioneerspacesim.net, go to [https://packages.ubuntu.com/xenial/libpng12-0](https://packages.ubuntu.com/xenial/libpng12-0).
Then extract (don't install) the Deb-Archive and copy the file `libpng12-0_1.2.54-1ubuntu1_amd64/lib/x86_64-linux-gnu/libpng12.so.0.54.0` into the folder with your Pioneer executable. After that, rename the copied file to "libpng12.so.0" that Pioneer can find it. The last thing you have to do is set the the `LD_LIBARY_PATH` environment variable so that the library search path is correct. Something like `LD_LIBRARY_PATH=. ./pioneer` will do when running from the folder the executable is in.
I hope that helps.

Still, it would be very good if the Linux build maintainers would adjust the libpng version. Why don't you include it in the build, anyways? Wouldn't that solve many compatibility issues?

Well, have fun with this awesome game & Cheers

**Comment by monolifed** [comment]
Then, shipping it with the library or using dynamic loading might help?

**Comment by Hoimar** [comment]
Yes, right, that would help much. But as far as I can see most people build Pioneer from source, that's why they don't have any issues. But for the normal user who wants to simply download and play the game without much hassle it's a very bad experience. And this issue exists since one full year now! I only was able to solve it because I'm a programmer and a bit of a linux nerd.

**Comment by achorein** [comment]
Thanks to Hoimar, I just add this on my DockerFile and it works great on stretch 
```
RUN wget http://fr.archive.ubuntu.com/ubuntu/pool/main/libp/libpng/libpng12-0_1.2.54-1ubuntu1_amd64.deb \
        && dpkg -x libpng12-0_1.2.54-1ubuntu1_amd64.deb /root/libpng12 \
        && ln -s /root/libpng12/lib/x86_64-linux-gnu/libpng12.so.0 /usr/lib/libpng12.so.0 \
        && ln -s /root/libpng12/lib/x86_64-linux-gnu/libpng12.so.0.54.0 /usr/lib/libpng12.so.0.54.0 \
        && ldconfig
```

**Comment by laarmen** [comment]
I'm confused. You made a dockerfile but you don't want to compile from
sources?

Le jeu. 29 mars 2018 à 11:37, Anselme Chorein <notifications@github.com> a
écrit :

**Comment by achorein** [comment]
Sorry, that was just about libpng12 on stretch, not about pionneers (apply on some nodejs build app).

**Comment by Hoimar** [comment]
Ah, thats why you replaced libpng12 globally.
I am still asking myself why so many projects do not ship their binary packages with already bundled libraries. I guess there`s a downside to that. But which?

**Comment by LuccoJ** [comment]
It's just not the way tree-based packaging systems, like the traditional ones in most Linux distributions, are meant to work.
It is the standard way to do it in things like Flatpak, though.

**Comment by Talkless** [comment]
How about using OBS, where multiple builds are provided for different distributions, as with this OwnCloud examle:
https://software.opensuse.org/package/owncloud

**Comment by sklei4** [comment]
You can download this onto Debian 9, and it will install the package. I'm working on a similar project that needs this library, so this should resolve your problem.
https://packages.ubuntu.com/trusty/amd64/libpng12-dev/download

**Comment by Ionis** [comment]
It might help.: `sudo apt-get install libpng-dev`

**Comment by FantasyNeurotic** [comment]
> Thanks to Hoimar, I just add this on my DockerFile and it works great on stretch
> 
> ```
> RUN wget http://fr.archive.ubuntu.com/ubuntu/pool/main/libp/libpng/libpng12-0_1.2.54-1ubuntu1_amd64.deb \
>         && dpkg -x libpng12-0_1.2.54-1ubuntu1_amd64.deb /root/libpng12 \
>         && ln -s /root/libpng12/lib/x86_64-linux-gnu/libpng12.so.0 /usr/lib/libpng12.so.0 \
>         && ln -s /root/libpng12/lib/x86_64-linux-gnu/libpng12.so.0.54.0 /usr/lib/libpng12.so.0.54.0 \
>         && ldconfig
> ```

Thanks,it is good for me.It solved the problem for me.

**Comment by se5a** [comment]
This is still a problem on linux Mint 19 which means it's likely a problem on ubuntu 18.

**Comment by exlipsed** [comment]
`sudo echo "deb http://mirrors.kernel.org/ubuntu/ xenial main" | sudo tee -a /etc/apt/sources.list && sudo apt-get update && sudo apt install -y --allow-unauthenticated libpng12-0`

should work in order to use libpng12-0 even on stretch - at least it worked on my stretch :)

**Comment by exlipsed** [comment]
yes it literally worked - I'm playing pioneer :) I actually solved this because the runescape-launcher had the same problem. But I'll give pioneer now a try :+1:

**Comment by zorigtt** [comment]
> Thanks to Hoimar, I just add this on my DockerFile and it works great on stretch
> 
> ```
> RUN wget http://fr.archive.ubuntu.com/ubuntu/pool/main/libp/libpng/libpng12-0_1.2.54-1ubuntu1_amd64.deb \
>         && dpkg -x libpng12-0_1.2.54-1ubuntu1_amd64.deb /root/libpng12 \
>         && ln -s /root/libpng12/lib/x86_64-linux-gnu/libpng12.so.0 /usr/lib/libpng12.so.0 \
>         && ln -s /root/libpng12/lib/x86_64-linux-gnu/libpng12.so.0.54.0 /usr/lib/libpng12.so.0.54.0 \
>         && ldconfig
> ```

Finally it worked. You saved me. Thank you so much.
